### ISSUE TITLE
Provide a schema validator

### ISSUE THREAD

**Issue by ben-manes** [jOOQ/jOOQ#2476]
It would be handy if jOOQ provided a schema validator so that an application can ensure that the generated code matches the runtime database schema. This is particularly useful when the code is generated from one database (such as H2 during for compile and test) and run against another (MySQL in production).

Due to the differences in the SQL dialects, we maintain two sets of flyway migration scripts to minimize confusion. However, that could lead to diverging schemas if a developer forget to update both sets of scripts. Most likely automated integration testing should catch this, but to better avoid surprises based on different side effects a validation phase at start up seems appropriate.

I currently have a [naive validator](https://gist.github.com/ben-manes/5638489), but I think a more intelligent one would be a useful part of jOOQ's tool chain.

**Comment by iceback** [comment]
I presume this is still not available?  I thought I had seen an update on this but cannot find same.

**Comment by lukaseder** [comment]
@iceback Nope. This seems quite out of scope / not well defined yet, to proceed. Any requirements / use cases you may want to share here are very welcome, of course.

**Comment by shorn** [comment]
Sharing a use-case:
I've been given a task to create an integration process that reads/updates an "integration table" in a database/schema we do not own.  The table structure will be defined and implemented by someone else.  I will be connecting to the DB with the integration table with JOOQ.  

My dev/testing will be done with a version of the table maintained by myself via Flyway scripts.  But when the process runs in integration-testing/production environments, I'm going to want to "sanity test" that integration table before kicking off the process.  

So, in this case, I think the proposal from Lukas that I want "a comparison and show differences" is suitable to what I want.  It's ok if the integration table is a little different from what I defined in development, as long as the SQL JOOQ executes can run.  So, I'm thinking I'd just want to see a "diff" report and decided for myself if the process should run or not.

Idea: it'd be pretty cool if I could give JOOQ my query and a DataSource/Connection and ask "hey, you reckon this query will run on this DB?"

**Comment by lukaseder** [comment]
Thanks for the ping, @shorn. Meanwhile, 7 years after issue creation and 1.5 years after my last comment, we actually have the infrastructure for such a feature: DDL interpretation and an actual diff API, which you can already use today:
https://www.jooq.org/doc/latest/manual/sql-building/schema-diff

I'm not convinced that we should implement a feature that checks if a given query *would run* on a given database. A much better approach to solving that problem is to implement an actual integration test on the given database version. It should be possible to re-create the production database version using your database change management (e.g. flyway), or via test containers, etc.

In the future, however, we'll be investing more time in this set of features via our new set of migrations features, where the above diff is a part of.

**Comment by tomdcc** [comment]
Another use-case that isn't supported by integration testing: we would like to support connecting our app to a database managed by customers, and it would be good to be able to verify that the schema is in an expected state.

Our migrations are defined in SQL rather than a more structured form such as a liquibase definition, so our generated jOOQ classes are the closest thing that we have to an inspectable expected schema. We can (and probably will) do a comparison manually by comparing database metadata to our jOOQ classes, but such a tool might be useful for others in similar scenarios.

**Comment by lukaseder** [comment]
> Another use-case that isn't supported by integration testing: we would like to support connecting our app to a database managed by customers, and it would be good to be able to verify that the schema is in an expected state.

These kinds of things are being worked on. I'm currently fixing a lot of issues in the current implementation of `DSLContext.meta()` to make this work. There's tooling around `Meta.migrateTo(Meta)` to compare two versions of schemas (e.g. from an actual connection to your DDL scripts), and it already works decently. So, if you'd like to run some tests and report some issues, that would be helpful!

> Our migrations are defined in SQL rather than a more structured form such as a liquibase definition, so our generated jOOQ classes are the closest thing that we have to an inspectable expected schema.

Or, just interpret the SQL using `DSLContext.meta(Source...)` and similar overloads?
